pygeo/DVConstraints.py

```diff
@@ -3133,6 +3484,336 @@ def writeTecplot(self, handle):
+            tmp_KS_p1 = self.DVGeo1.totalSensitivity(np.transpose(deriv_outputs[6]), self.surface_1_name+'_p1', config=config)
+            tmp_KS_p2 = self.DVGeo1.totalSensitivity(np.transpose(deriv_outputs[7]), self.surface_1_name+'_p2', config=config)
+            tmp_perim_p0 = self.DVGeo1.totalSensitivity(np.transpose(deriv_outputs[11]), self.surface_1_name+'_p0', config=config)
+            tmp_perim_p1 = self.DVGeo1.totalSensitivity(np.transpose(deriv_outputs[12]), self.surface_1_name+'_p1', config=config)
+            tmp_perim_p2 = self.DVGeo1.totalSensitivity(np.transpose(deriv_outputs[13]), self.surface_1_name+'_p2', config=config)
+            for key in tmp_KS_p0:
+                tmpTotalKS[key] = tmp_KS_p0[key]+tmp_KS_p1[key]+tmp_KS_p2[key]
+                tmpTotalPerim[key] = tmp_perim_p0[key]+tmp_perim_p1[key]+tmp_perim_p2[key]
+
+
+        if self.DVGeo2 is not None:
+            nDV2 = self.DVGeo2.getNDV()
+        else:
+            nDV2 = 0
+
+        if nDV2 > 0:
+            # compute sensitivity with respect to the first mesh
+            # grad indices 0-2 are for mesh 1 p0, 1, 2 / 3-5 are for mesh 2 p0, 1, 2
+            tmp_KS_p0 = self.DVGeo2.totalSensitivity(np.transpose(deriv_outputs[8]), self.surface_2_name+'_p0', config=config)
+            tmp_KS_p1 = self.DVGeo2.totalSensitivity(np.transpose(deriv_outputs[9]), self.surface_2_name+'_p1', config=config)
+            tmp_KS_p2 = self.DVGeo2.totalSensitivity(np.transpose(deriv_outputs[10]), self.surface_2_name+'_p2', config=config)
+            tmp_perim_p0 = self.DVGeo2.totalSensitivity(np.transpose(deriv_outputs[14]), self.surface_2_name+'_p0', config=config)
+            tmp_perim_p1 = self.DVGeo2.totalSensitivity(np.transpose(deriv_outputs[15]), self.surface_2_name+'_p1', config=config)
+            tmp_perim_p2 = self.DVGeo2.totalSensitivity(np.transpose(deriv_outputs[16]), self.surface_2_name+'_p2', config=config)
+            for key in tmp_KS_p0:
+                tmpTotalKS[key] = tmp_KS_p0[key]+tmp_KS_p1[key]+tmp_KS_p2[key]
+                tmpTotalPerim[key] = tmp_perim_p0[key]+tmp_perim_p1[key]+tmp_perim_p2[key]
+        funcsSens[self.name+'_KS'] = tmpTotalKS
+        funcsSens[self.name+'_perim'] = tmpTotalPerim
+
+    def evalTriangulatedSurfConstraint(self):
+        """
+        Call geograd to compute the KS function and intersection length
+        """
+        # first compute the length of the intersection surface between the object and surf mesh
+        from geograd import geograd_parallel
+        mindist_tmp = 0.0
+
+        # first run to get the minimum distance
+        foo, perim_length, mindist, foo2, foo3 = geograd_parallel.compute(self.surf1_p0, self.surf1_p1, self.surf1_p2,
+                self.surf2_p0, self.surf2_p1, self.surf2_p2, mindist_tmp, self.rho, self.maxdim)
+        # second run gets the well-conditioned KS
```

Comment:
I don't follow. Why does the KS value change between runs?

Reply:
Should these be separate functions?

Reply:
It changes because to obtain a well-conditioned computation of the KS function you subtract off the current minimum value. In exact math it doesn't matter but with computer math it does. So you feed the min dist from the prior run into the second run to get the "real
 KS value

Reply:
Ok that makes sense. 

Could these two functional calls be combined within geograd so it never returns and ill-conditioned KS?

Reply:
Returning values that aren't accurate or shouldn't be used is a bad idea in my opinion.

@anilyil  thoughts?

Reply:
@joanibal combining the calls isn't practical because the number of pairwise comparisons is too large to hold in memory pending the minimum being found. geograd isn't intended to be a public API so IMO the way it is now is OK practice (I edited the foos to underscores to make the intent clear)

Reply:
OKk, I suppose that's fine